Libs/vtkSegmentationCore/vtkOrientedImageDataResample.cxx

```diff
@@ -1835,9 +1835,100 @@ bool vtkOrientedImageDataResample::IsLabelInMask(
   return valueFound;
 }
 
+//----------------------------------------------------------------------------
+bool vtkOrientedImageDataResample::ValidateSegmentationImageType(vtkImageData* image)
+{
+  if (!image)
+  {
+    vtkErrorWithObjectMacro(nullptr, "ValidateSegmentationImageType: Invalid input image");
+    return false;
+  }
+
+  vtkIdType validScalarTypes[3] =
+  {
+    VTK_UNSIGNED_CHAR,
+    VTK_UNSIGNED_SHORT,
+    VTK_UNSIGNED_INT
+    //VTK_UNSIGNED_LONG // See issue #5427.
+                        // vtkImageThreshold is not able to fill thresholded images with VTK_LONG_MAX, preventing segment erase.
+  };
+
+  int scalarType = image->GetScalarType();
+  for (int i = 0; i < 3; i++)
+  {
+    if (scalarType == validScalarTypes[i])
+    {
+      return true;
+    }
+  }
+
+  double scalarRange[2] = { -1.0, 0.0 };
+  image->GetScalarRange(scalarRange);
+  if (scalarRange[0] < 0)
+  {
+    vtkErrorWithObjectMacro(image, "ValidateSegmentationImageType: Negative values are present, so image cannot be cast to unsigned type");
+    return false;
+  }
+
+  vtkGenericWarningMacro("ValidateSegmentationImageType: Invalid scalar type (" << image->GetScalarTypeAsString()
+    << "), converting image to smallest unsigned integer type");
+
+  // Attempt to cast the image to a valid scalar type.
+  return vtkOrientedImageDataResample::CastToSmallestUnsignedIntegerType(image);
+}
 
 //----------------------------------------------------------------------------
-void vtkOrientedImageDataResample::CastImageForValue(vtkOrientedImageData* image, double value)
+bool vtkOrientedImageDataResample::CastToSmallestUnsignedIntegerType(vtkImageData* image)
+{
+  if (!image)
+  {
+    vtkErrorWithObjectMacro(nullptr, "CastImageToSmallestIntegerType: Invalid input image");
+    return false;
+  }
+
+  double scalarRange[2];
+  image->GetScalarRange(scalarRange);
+  if (scalarRange[0] < 0)
+  {
+    vtkErrorWithObjectMacro(image, "CastImageToSmallestIntegerType: Negative values are present, so image cannot be cast to unsigned type");
+    return false;
+  }
+
+  int scalarType = image->GetScalarType();
+  if (scalarRange[1] <= VTK_UNSIGNED_CHAR_MAX)
+  {
+    scalarType = VTK_UNSIGNED_CHAR;
+  }
+  else if (scalarRange[1] <= VTK_UNSIGNED_SHORT_MAX)
+  {
+    scalarType = VTK_UNSIGNED_SHORT;
+  }
+  else if (scalarRange[1] <= VTK_UNSIGNED_INT_MAX)
+  {
+    scalarType = VTK_UNSIGNED_INT;
+  }
+  else
+  {
+    vtkErrorWithObjectMacro(image, "CastImageToSmallestIntegerType: Scalar range is too large to fit into an integer type");
+    return false;
+  }
+
+  if (scalarType == image->GetScalarType())
```

Comment:
Shouldn't we also check that any floating point data is losslessly castable to integer and warn if not?  Otherwise we would be silently losing data.  I know at least some of the float data that a user tried to load as a segmentation recently had non-integral values.

Reply:
We don't check specifically for lossless castability, however we always log a warning when casting from a floating point type to an integer type so that the user knows that it is a possibility.

The non-integral values are truncated during cast, which is consistent with how they are displayed in labelmap nodes.